kaolin/datasets/base.py

```diff
@@ -0,0 +1,154 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from tqdm import tqdm
+
+import torch
+from torch.multiprocessing import Pool
+from torch.utils.data import Dataset
+
+from kaolin import helpers
+
+def _preprocess_task(args):
+    torch.set_num_threads(1)
+    with torch.no_grad():
+        idx, get_data, get_attributes, cache_transform = args
+        name = get_attributes(idx)['name']
+        if name not in cache_transform.cached_ids:
+            data = get_data(idx)
+            cache_transform(name, *data)
+
+class KaolinDatasetMeta(type):
+    def __new__(metacls, cls_name, base_cls, class_dict):
+        if cls_name != "KaolinDataset":
+            class_dict['__doc__'] += \
+                """Additional args:
+        preprocessing_params (dict): parameters for the preprocessing:
+            - 'cache_dir': path to the cached preprocessed data.
+            - 'num_workers': number of process used in parallel for preprocessing (default: number of cores)
+        preprocessing_transform (Callable): Called on the outputs of _get_data over the indices
+                                            from 0 to len(self) during the construction of the dataset,
+                                            the preprocessed outputs are then cached to 'cache_dir'.
+        transform (Callable): Called on the preprocessed data at __getitem__.
+        no_progress (bool): disable tqdm progress bar for preprocessing."""
+        return type.__new__(metacls, cls_name, base_cls, class_dict)
+
+class KaolinDataset(Dataset, metaclass=KaolinDatasetMeta):
+    """
+    Abstract class for dataset with handling of multiprocess or cuda preprocessing.
+
+    A KaolinDataset children class will need the above implementation:
+       1) _initialize:
+           Initialization function called at the beginning of the constructor.
+       2) _get_data:
+           Data getter that will be preprocessed => cached => transformed, take an index as input.
+       3) _get_attributes:
+           Attributes getter that will be preprocess / transform independent.
+       4) __len__:
+           Return the size of the dataset
+    """
+    def __init__(self, *args, preprocessing_transform=None, preprocessing_params: dict = None,
+                 transform=None, no_progress: bool = False, **kwargs):
+        """
+        Args:
+            positional and keyword arguments for initialize(*args, **kwargs) (see class and initialize documentation)
```

Comment:
Is there a way we can "link" from here to the metaclass docs?

Reply:
It leads to indentations problems, also please notice that the docstrings are not exactly the same

Reply:
Ah I see. I was thinking of the way pytorch does it. Eg. https://pytorch.org/docs/stable/tensors.html#torch.Tensor.sigmoid

Not a concern atm, was just wondering.